### PR TITLE
Remove v4.1+ deprecated code from modeling

### PR DESCRIPTION
**Pull request by WilliamJamieson**
### Description

The following has been marked as deprecated (with removal starting in v4.1) in modeling:

- [`ExpressionTree`](https://github.com/astropy/astropy/blob/4ae21dadab159f7e6bc8157f059237fec14baead/astropy/modeling/utils.py#L22-L295)
- [`MexicanHat1D` and `MexicanHat2D`](https://github.com/astropy/astropy/blob/4ae21dadab159f7e6bc8157f059237fec14baead/astropy/modeling/functional_models.py#L2838-L2845)
- [`inputs` in `Model`](https://github.com/astropy/astropy/blob/4ae21dadab159f7e6bc8157f059237fec14baead/astropy/modeling/core.py#L722-L733)
- [`CompoundModel` `inverse` parameter](https://github.com/astropy/astropy/blob/4ae21dadab159f7e6bc8157f059237fec14baead/astropy/modeling/core.py#L2536-L2542)
- [`CompoundModel.both_inverses_exist()` method](https://github.com/astropy/astropy/blob/4ae21dadab159f7e6bc8157f059237fec14baead/astropy/modeling/core.py#L2788-L2804)

This PR removes all of these things for future astropy versions.

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [x] Do the proposed changes actually accomplish desired goals?
- [x] Do the proposed changes follow the [Astropy coding guidelines](https://docs.astropy.org/en/latest/development/codeguide.html)?
- [x] Are tests added/updated as required? If so, do they follow the [Astropy testing guidelines](https://docs.astropy.org/en/latest/development/testguide.html)?
- [ ] Are docs added/updated as required? If so, do they follow the [Astropy documentation guidelines](https://docs.astropy.org/en/latest/development/docguide.html#astropy-documentation-rules-and-guidelines)?
- [ ] Is rebase and/or squash necessary? If so, please provide the author with appropriate instructions. Also see ["When to rebase and squash commits"](https://docs.astropy.org/en/latest/development/when_to_rebase.html).
- [x] Did the CI pass? If no, are the failures related? If you need to run daily and weekly cron jobs as part of the PR, please apply the `Extra CI` label.
- [x] Is a change log needed? If yes, did the change log check pass? If no, add the `no-changelog-entry-needed` label.
- [ ] Is a milestone set? Milestone must be set but `astropy-bot` check might be missing; do not let the green checkmark fool you.
- [ ] At the time of adding the milestone, if the milestone set requires a backport to release branch(es), apply the appropriate `backport-X.Y.x` label(s) *before* merge.

### CODE REVIEW

**Comment by WilliamJamieson** [comment]
Tagging @nden and @perrygreenfield.

**Comment by pllim** [comment]
If I remember correctly in a recent discussion `docs/changes/modeling/11978.other.rst` should really be in `docs/changes/11978.other.rst`, or you should rename to `docs/changes/modeling/11978.api.rst`? I am always confused by `towncrier`, so maybe @saimn or @eteq can clarify.

**Comment by pllim** [comment]
p.s. There were also discussions on max line length (5 lines?) for the change log fragment.

**Comment by WilliamJamieson** [comment]
> p.s. There were also discussions on max line length (5 lines?) for the change log fragment.

It is difficult to read if I reduce the text to fewer lines in my opinion. However, I can do so if this is what is desired now.

File: docs/changes/modeling/11978.api.rst [hunk 1]
```diff
@@ -0,0 +1,7 @@
+Removed the following deprecated modeling features:
+    ``~astropy.modeling.utils.ExpressionTree`` class,
```
Comment (bsipocz) [inline docs/changes/modeling/11978.api.rst:2]: extreme minor nit-pick: remove the `~` as these are not rendered to sphinx links. otoh I think it's good to keep their full API path in the changelog, even when they may be available from a different namespace, too.

**Comment by pllim** [comment]
Re: change log length -- you can ignore me. 5 line isn't hard limit. 😅

File: docs/changes/modeling/11978.api.rst (hunk 1 above)
Comment (WilliamJamieson) [reply docs/changes/modeling/11978.api.rst]: See ed99fdd

**Review by nden — APPROVED** [review]

**Pull request merged**